src/analysis/souffle/taint.dl

```diff
@@ -110,26 +110,68 @@ mayHaveTag(tgt, tag) :-
 // These rules add objects to the set of things that exist (i.e. relations like
 // `isObject(..)`) whenever they are mentioned in other rules.
 
-// Symbols used in edges are access paths
-isAccessPath(x) :- edge(x, _).
-isAccessPath(y) :- edge(_, y).
+// Access paths directly mentioned by other facts.
+.decl isDirectlyMentionedAccessPath(accessPath: AccessPath)
 
-isAccessPath(path) :- claimHasTag(path, _).
-isAccessPath(path) :- mayHaveTag(path, _).
+// Symbols used in edges are access paths
+isDirectlyMentionedAccessPath(x) :- edge(x, _).
+isDirectlyMentionedAccessPath(y) :- edge(_, y).
+
+isDirectlyMentionedAccessPath(path) :- claimHasTag(path, _).
+isDirectlyMentionedAccessPath(path) :- mayHaveTag(path, _).
+
+// Enumerates all potentially-valid indices for any directly-mentioned access path. Because all
+// derived access paths are substrings of directly-mentioned access paths, this gives all
+// potentially-valid indices for all access paths.
+.decl isAccessPathIndex(index: number)
+isAccessPathIndex(0).
+isAccessPathIndex(index + 1) :-
+  isAccessPathIndex(index),
+  isDirectlyMentionedAccessPath(accessPath),
+  index + 1 < strlen(accessPath).
+
+// Access paths split into subpaths at the dot character; this finds the locations at which that
+// split occurs for each access path.
+.decl isMemberBoundary(accessPath: AccessPath, index: number)
+isMemberBoundary(accessPath, index) :-
+  isDirectlyMentionedAccessPath(accessPath),
+  isAccessPathIndex(index),
+  index < strlen(accessPath),
+  substr(accessPath, index, 1) = ".".
+
+// Any access path that was directly mentioned is an access path.
+isAccessPath(accessPath) :- isDirectlyMentionedAccessPath(accessPath).
+
+// Any prefix of a directly mentioned access path that ends in a dot is an access path (not
+// including the dot).
+isAccessPath(substr(accessPath, 0, boundaryIdx)) :-
```

Comment:
Datalog computes all the relations eagerly. Therefore, this rule forces datalog solver to compute tags for all partial prefixes. The advantage of defining `isAccessPath` from `edge`, `claimHasTag`, `mayHaveTag`, etc., is that we will only compute tags for the access paths that we care about. 

Is this new inference rule needed for correctness of the `isMemberOf` rule?

Reply:
I think, unfortunately, that either this elaboration or an alternative mechanism is needed to accomplish the `isMemberOf` rule. See the `isMemberOf` rule below:

```
isMemberOf(base, member) :-
  isAccessPath(base),
  isAccessPath(member),
  strlen(base) + 1 < strlen(member),
  cat(base, ".") = substr(member, 0, strlen(base) + 1).
  ```

In order to ground the variables used in the rule, I have to check that they are within `isAccessPath`. Here are the rules that populate `isAccessPath` other than those that consider partial prefixes:

```
// Access paths directly mentioned by other facts.
.decl isDirectlyMentionedAccessPath(accessPath: AccessPath)

// Symbols used in edges are access paths
isDirectlyMentionedAccessPath(x) :- edge(x, _).
isDirectlyMentionedAccessPath(y) :- edge(_, y).

isDirectlyMentionedAccessPath(path) :- claimHasTag(path, _).
isDirectlyMentionedAccessPath(path) :- mayHaveTag(path, _).

// Any access path that was directly mentioned is an access path.
isAccessPath(accessPath) :- isDirectlyMentionedAccessPath(accessPath).
```

If `member` is a leaf, it will be directly mentioned by an `edge` or `claim`, so that is not a problem. If `base` is an inner node, however, (as we expect it to be in all cases that are not trivially false) it will not be mentioned by an `edge`. It will also not be the subject of a `claimHasTag` (even if it were, this would be a trivially-true case, and not very interesting). The only way that an inner node can become part of `isAccessPath` is through `mayHaveTag`.

However, the one rule that could place it in `mayHaveTag` is the rule that propagates tags from members to bases. Therefore, there is a circular dependency unless we get this partial path into `isAccessPath` in a manner that is more directly connected to base facts.

If we do not want to derive all possible partial prefixes, I think the best way to do this is to introduce a new kind of fact that indicates what paths are of interest to checks. For instance, we could use the following:

```
.decl checkedAccessPath(accessPath: AccessPath)

isAccessPath(accessPath) :- checkedAccessPath(accessPath).
```

and introduce a `checkedAccessPath` for all inner nodes that are the subject of a check. A real system, of course, would probably just introduce one of these for any check node, but we might as well save ourselves some typing.

This is similar to the idea I had initially of introducing nodes for `checkHasTag` and friends, but unlike those nodes, this kicks the question of "how do we represent complex predicates on tags" down the road. We simply indicate that an access path is checked and allow that to populate the universe of `AccessPath`s.

What do you think?

Reply:
> If we do not want to derive all possible partial prefixes, I think the best way to do this is to introduce a new kind of fact that indicates what paths are of interest to checks. For instance, we could use the following:
> ```
> .decl checkedAccessPath(accessPath: AccessPath)
> isAccessPath(accessPath) :- checkedAccessPath(accessPath).
> ```
> and introduce a checkedAccessPath for all inner nodes that are the subject of a check. A real system, of course,  would probably just introduce one of these for any check node, but we might as well save ourselves some typing

I was thinking of something along these lines. If we go down this route, will it be sound if accidentally miss a  `isAccessPath` predicate for an access path?

Reply:
When we get to the point where we're automatically generating these rules, I think we're helped out by the fact that `check`s are structures explicitly given to us by our IR; it will be hard to forget to generate these check facts (or some successor to them containing a predicate). While we're hand-writing tests, though, I think it is valid to be concerned about forgetting to drop in these checks for non-leaf nodes. I can't immediately think of a good way to force us to not miss them, but I'll think on it.

Reply:
Perhaps, the simplest is to also do verify that `isAccessPath` is true as a part of the test case. We could force this by defining a macro?

Reply:
I've added `CHECK_TAG_PRESENT` and `CHECK_TAG_NOT_PRESENT` to do this.